Token based authentication

19ec94 opened the pull request:
This Pull Request will
 - re-organise project structure
 - refactor router logic into single-purpose  middleware components 
 - check user input validity 
 - verify log in or sign up
 - generate JWT token
 - send token to client
 - authenticate JWT token validity.

Rationale for proposed changes:
 - to promote modular development
 - to make middleware reusable in multiple routes, thus reducing duplicity

19ec94 committed:
Install jsonwebtoken library

19ec94 committed:
Modify the root path for user management

19ec94 committed:
Update the comment

19ec94 committed:
Add logic for validating data input

Rationale:
 Introduction of the userValidator early in the middleware chain ensures
 server does not perfrom any additional service if the user input does
 not meet the criteria

19ec94 committed:
Refer the commit message in 4ba80e35f400bcabad661af151805f0d4d6f0ab3

19ec94 committed:
Add support for token creation upon signup/login

This commit will
 - add jsonwebtoken library
 - create a token after sucessful signup or login

19ec94 committed:
Create functionality for automatic login upon successful signup

19ec94 committed:
Modify response JSON API to include data/error message

19ec94 committed:
Replace hard-coded values by environment variables

19ec94 committed:
Move files and subdirectories from  dir to project root bin	     docker-compose.yml  models        package-lock.json  routes
config	     index.js		 node_modules  public		  views
controllers  middlewares	 package.json  README.md

19ec94 committed:
Add entry point for the server

19ec94 committed:
Rename file from index.js to app.js

19ec94 committed:
Fix typo

19ec94 committed:
Move server setup to bin/www

This commit will also:
 - Reorganise code structure
 - Fix import modules statements
 - Add helpful comments
 - Replace single quotes by double quotes

19ec94 committed:
Modify entry point for server

19ec94 committed:
Add logic to get root/index ('/') page

19ec94 committed:
Change file name for consistency sake

19ec94 committed:
Fix import statement

19ec94 committed:
Format file content

19ec94 committed:
Remove unnecessary dotenv import statement

19ec94 committed:
Change variable name to js-style for consitency

19ec94 committed:
Add comments

19ec94 committed:
Delete userValidator file

Reason for deletion:
 userValidator file is a collection of several pieces of logic to validate the
 user input data: signup data, login data etc
Alternative suggestion:
 - split each logic into its own file - isLoginDataValid.middleware.js,
   isSignupDataValid.middleware.js, etc
 - collect all the related files in a index file - e.g
   user.middleware.index.js
 - import this index to access each middleware logic

19ec94 committed:
Add logic for validating user login/sigup input

See the commit message in f00a574e5b5d1b5c7c8230f6b338cf104473a23f

19ec94 committed:
Fix import statements and object referencing

19ec94 committed:
Rename file for consistency

Naming convention:
 <descriptiveName or GroupName>.<technicalName>(.index).js

19ec94 committed:
Fix import statements

Read the message in commit 728051e8289f0078fa41b84b5c35c27798e3e4b7

19ec94 committed:
Add comments

19ec94 committed:
Fix import statements

19ec94 committed:
Add logic to endpoint controller functions

See commit message f00a574e5b5d1b5c7c8230f6b338cf104473a23f

19ec94 committed:
Add all user related controller endpoints

19ec94 committed:
Fix import statements and variable accessing method

19ec94 committed:
Remove controller logic for signup

19ec94 committed:
Simply controller logic

This commit will
 - split the sendSendSession token into several middleware functions
   which will
   - check is and existing user,
   - verify password,
   - generate session token,
 - and then finally send the session token to client.

 This split opens up modularity, reusablity, extendability.

19ec94 committed:
Add middlewares isNewUser and addNewUserToDatabase for signup route

19ec94 committed:
Remove the old controller logic for addUserToDatabase

19ec94 committed:
Add several middleware logic

This commit will add the following logic to
 - check whether the username exists in database,
 - verify password,
 - generate new session token,
 - send session token to client
in the `/login` route.

This commit will also add the following logic to
 - check whether the username is new,
 - add new user to the database
 - generate new session token,
 - send session token to client
in the `/signup` route.

19ec94 committed:
Add comments

19ec94 committed:
Remove unnecessary email variable

19ec94 committed:
Fix error message format

19ec94 committed:
Remove default values in database setup

19ec94 committed:
Add middleware to authenticate session token

19ec94 committed:
Add attributes to cookies

19ec94 committed:
Remove debugging statement

rmorasis commented:
The refactored structure of the project looks good to me and makes sense.
I did run the server and did some manual tests of the endpoints. All seems to work as intended.
Some minor suggestions for improvement that do not necessarily need to happen in this merge request are:

- Add a small section about “how to run the server” to the README

- Add some documentation for the required/expected environmental variables either to the README or even better as .env file without values (just the names and then gitignore will protect us from accidentally committing values):

```
DB_HOST=
DB_PORT=
DB_NAME=
...
```

- Maybe improve validation of the payloads since sending no payload will currently crash the server (validation could be taken care of with express-openapi, see next point)

- This might be a bit overkill for this project, but adding OpenAPI documentation and validation via express-openapi would be nicer and more friendly to use

rmorasis reviewed (commented):
The refactored structure of the project looks good to me and makes sense.
I did run the server and did some manual tests of the endpoints. All seems to work as intended.
Some minor suggestions for improvement that do not necessarily need to happen in this merge request are:

- Add a small section about “how to run the server” to the README

- Add some documentation for the required/expected environmental variables either to the README or even better as .env file without values (just the names and then gitignore will protect us from accidentally committing values):

```
DB_HOST=
DB_PORT=
DB_NAME=
...
```

- Maybe improve validation of the payloads since sending no payload will currently crash the server (validation could be taken care of with express-openapi, see next point)

- This might be a bit overkill for this project, but adding OpenAPI documentation and validation via express-openapi would be nicer and more friendly to use

19ec94 committed:
Update readme file

Add instructions
 - to notify users of expected environment variables,
 - on using the server and database

19ec94 commented:
@rmorasis  Thanks for the comments. I want to raise two points:
 1. I updated the README file to reflect the first two points in your comment above. Please could you verify it by following the instructions on the sections "environment variable" and "running the server"?
 2. I agree that more to be done on the validation of user input. I only created a skeleton and a basic mechanism. But I would like to reproduce the server crash you mentioned. Could you please share the endpoint and the payload you used to crash the server?
 
 If you feel the recent changes are satisfactory, approve this PR.

19ec94 committed:
Fix typo

rmorasis commented on README.md:
```diff
@@ -61,3 +61,52 @@ http://localhost:5000/username
+ - DB_NAME=i3wd
+ - DB_USER=root
+ - DB_PASSWORD=rootpassword
+ - PORT_BACKEND=5000
```
As far as I can see in "bin/www" it´s expecting "PORT" and not not "PORT_BACKEND"
```suggestion
 - PORT=5000
```

rmorasis commented on README.md:
````diff
@@ -61,3 +61,52 @@ http://localhost:5000/username
+
+  3. To establish an interactive connection to the database using mongosh, execute
+  ```bash
+  sudo docker exec -it mongodb mongosh -u root -p rootpassword --authenticationDatabase admin 
````
Nit: It´s consistent in the documentation, but maybe is not obvious that user and password need to correspond to the values of the environment variables DB_USER and DB_PASSWORD.
```suggestion
  sudo docker exec -it mongodb mongosh -u <DB_USER> -p <DB_PASSWORD> --authenticationDatabase admin 
```

rmorasis reviewed (commented):
@19ec94 the changes look great. just 2 small nitpicks/typos. Otherwise I´m happy to approve.

19ec94 committed:
Fix typo and update info

19ec94 commented:
@rmorasis  Yeah, you are right about both points. I fixed the typo and updated the info on running the server.

19ec94 commented:
@rmorasis  Thanks for the swift review and approval. Let me close this PR and delete the branch `token-based-authentication`.
